Emby.Server.Implementations/ScheduledTasks/Tasks/CleanupCollectionPathsTask.cs

```diff
@@ -0,0 +1,117 @@
+using MediaBrowser.Model.IO;
+using MediaBrowser.Model.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Emby.Server.Implementations.ScheduledTasks.Tasks;
+
+/// <summary>
+/// Deletes Path references from collections that no longer exists.
+/// </summary>
+public class CleanupCollectionPathsTask : IScheduledTask
+{
+    private readonly ILocalizationManager _localization;
+    private readonly ICollectionManager _collectionManager;
+    private readonly ILogger<CleanupCollectionPathsTask> _logger;
+    private readonly IProviderManager _providerManager;
+    private readonly IFileSystem _fileSystem;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="CleanupCollectionPathsTask"/> class.
+    /// </summary>
+    /// <param name="localization">Instance of the <see cref="ILocalizationManager"/> interface.</param>
+    /// <param name="collectionManager">Instance of the <see cref="ICollectionManager"/> interface.</param>
+    /// <param name="logger">The logger.</param>
+    /// <param name="providerManager">The provider manager.</param>
+    /// <param name="fileSystem">The filesystem.</param>
+    public CleanupCollectionPathsTask(
+        ILocalizationManager localization,
+        ICollectionManager collectionManager,
+        ILogger<CleanupCollectionPathsTask> logger,
+        IProviderManager providerManager,
+        IFileSystem fileSystem)
+    {
+        _localization = localization;
+        _collectionManager = collectionManager;
+        _logger = logger;
+        _providerManager = providerManager;
+        _fileSystem = fileSystem;
+    }
+
+    /// <inheritdoc />
+    public string Name => _localization.GetLocalizedString("TaskCleanCollections");
+
+    /// <inheritdoc />
+    public string Key => "CleanCollections";
+
+    /// <inheritdoc />
+    public string Description => _localization.GetLocalizedString("TaskCleanCollectionsDescription");
+
+    /// <inheritdoc />
+    public string Category => _localization.GetLocalizedString("TasksMaintenanceCategory");
+
+    /// <inheritdoc />
+    public async Task ExecuteAsync(IProgress<double> progress, CancellationToken cancellationToken)
+    {
+        var collectionsFolder = await _collectionManager.GetCollectionsFolder(false).ConfigureAwait(false);
+        if (collectionsFolder is null)
+        {
+            _logger.LogInformation("There is no collection folder to be found.");
+            return;
+        }
+
+        var collections = collectionsFolder.Children.OfType<BoxSet>().ToArray();
+        _logger.LogTrace("Found {CollectionLength} Boxsets.", collections.Length);
+
+        for (var index = 0; index < collections.Length; index++)
+        {
+            var collection = collections[index];
+            _logger.LogTrace("Check Boxset {CollectionName}.", collection.Name);
+            var itemsToRemove = new List<LinkedChild>();
+            foreach (var collectionLinkedChild in collection.LinkedChildren.ToArray())
```

Comment:
```suggestion
            foreach (var collectionLinkedChild in collection.LinkedChildren)
```

this is already an array

Reply:
As this seems to be a shared object, i want to limit the possibility on a concurrent access to that collection. If you think this is unnecessary i will remove the enumeration.

Reply:
Yes, we should remove calling .ToArray() as it won't do anything, except that it will copy the array items to a new array. `LinkedChildren` is reassigned to different arrays in other parts of the code, however which items are in the instance are never altered at any point. (i.e. we only ever do `LinkedChildren = ...`, not `LinkedChildren[3] = ...`).

Because an invocation of `foreach` like this:

```
foreach (var collectionLinkedChild in collection.LinkedChildren)
{
    // ...
}
```

Is, for all intents and purposes, just syntactic sugar for this:

```
var enumerator = collection.LinkedChildren.GetEnumerator();
while (enumerator.MoveNext())
{
    var collectionLinkedChild = enumerator.Current;   
    // ...
}
```

That means that even if the original array reference is assigned to a different array during execution, once the foreach loop starts executing, it will finish executing on the same array it started on, rather than switching to the new array partway through. 

Note: this is not true for `for`: 

```
for (int i = 0; i < collection.LinkedChildren; i++) 
{
    var collectionLinkedChild = collection.LinkedChildren[i];
    // ...
}
```

In this case, if the array was reassigned to a different array, it would actually switch to the new array reference, rather than complete execution on the original array, due to the fact that the array property is actually referenced on each iteration of the array. 

P.S. The compiler runs on eldritch horror and programmer tears, so it might decide to compile things differently due to deep magic if it's more efficient, but this is define behavior, so you can rely on it acting the same, even if the underlying execution is different.

Reply:
That is more to prevent modifications on that enumerator to cause this to throw
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L1181

But i digress.